index.js

```diff
@@ -1,75 +1,162 @@
 'use strict'
 
-var http = require('http')
-var parseUrl = require('url').parse
-var zlib = require('zlib')
-var stringify = require('fast-safe-stringify')
-var pkg = require('./package')
+const parseUrl = require('url').parse
+const zlib = require('zlib')
+const pump = require('pump')
+const ndjson = require('ndjson')
+const safeStringify = require('fast-safe-stringify')
+const streamToBuffer = require('fast-stream-to-buffer')
+const StreamChopper = require('stream-chopper')
+const pkg = require('./package')
+
+module.exports = Client
+
+function Client (opts, onerror) {
+  opts = normalizeOptions(opts)
+  onerror = onerror || throwerr
+
+  const serialize = ndjson.serialize()
+  // TODO: Is there a race condition between the `time` config option and the request timeout option?
+  const chopper = new StreamChopper(opts).on('stream', onStream(opts, onerror))
+
+  // All sockets on the agent are unreffed when they are created. This means
+  // that when those are the only handles left, the `beforeExit` event will be
+  // emitted. By listening for this we can make sure to end the requests
+  // properly before exiting. This way we don't keep the process running until
+  // the `time` timeout happens.
+  // TODO: If you start too many clients (e.g. in a test), this will emit a too-many-listeners warning
```

Comment:
You could stuff the clients in an array and only attach this handle once at the global scope, iterating over the clients and running a destructor of sorts to do this cleanup.

Reply:
Much better idea. Fixed 👍